### PR TITLE
[model, rollout, cfg, tests, doc] feat: Qwen3-Omni Thinker LoRA FSDP on transformers 5.x

### PR DESCRIPTION
**Pull request by qinganrice**
### What does this PR do?

Stacked on #211,  will rebase onto main after it merges

Adds **transformers 5.x** support to the Qwen3-Omni Thinker GSPO + LoRA recipe so the full MoE FFN (expert `gate_proj` / `up_proj` / `down_proj`) is LoRA-trainable on both 4.x and 5.x.

transformers 5.x stores the Thinker MoE experts as fused 3-D tensors (`gate_up_proj` / `down_proj` `nn.Parameter`), which PEFT cannot target for LoRA. This PR hooks `peft.get_peft_model` to unfuse those experts into per-expert `nn.Linear` right before LoRA injection (weights cloned from the fused params, forward semantics preserved), and no-ops PEFT 5.x's `gate_up_proj` name-remapping for Qwen3-Omni. On 4.x the experts are already `nn.Linear`, so the same `target_modules` config works unchanged.

Follow-up to the base Qwen3-Omni Thinker recipe (already merged); builds on latest `main`.

Related issues / PRs: https://github.com/verl-project/verl-omni/pull/113

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here: <FILL: e.g. https://github.com/verl-project/verl-omni/pulls?q=qwen3-omni+thinker>
- [x] Format the PR title as `[{modules}] {type}: {description}`
  - `[model, rollout, cfg, tests, doc] feat: Qwen3-Omni Thinker MoE LoRA on transformers 5.x`

### Test

Validated end-to-end with the GSPO + LoRA smoke test (tiny random Qwen3-Omni, 2 steps, vLLM-Omni AR rollout, FSDP LoRA sync) and full model test:

Result: training completes both steps; rollout↔actor logprob `pearson ≈ 0.9998`. Verified PEFT targets the unfused MoE experts (48 expert LoRA adapters: 4 experts × {gate, up, down} × 2 layers × {A, B}). Re-added to `gpu_smoke` as **Test 8** (2 GPUs).

Full model training reward results (val is 0.887):
Screenshot 2026-06-26 at 12 26 22 PM

### API and Usage Example

No API change. The recipe config now lists the MoE FFN modules explicitly (`all-linear` cannot resolve the unfused experts on the multimodal wrapper):

```bash
actor_rollout_ref.model.target_modules="q_proj,k_proj,v_proj,o_proj,gate_proj,up_proj,down_proj"
actor_rollout_ref.actor.fsdp_config.use_orig_params=True
actor_rollout_ref.actor.fsdp_config.wrap_policy.min_num_params=100000000
```

### Design & Code Changes

- `verl_omni/models/transformers/qwen3_omni_thinker.py` — add `_patch_unfuse_qwen3_omni_thinker_experts` (tf5-only, idempotent): unfuse experts into per-expert `nn.Linear` in a `get_peft_model` hook; no-op PEFT 5.x `convert_peft_config_for_transformers` remapping for Qwen3-Omni; split comma-string `target_modules`.
- `verl_omni/workers/rollout/vllm_rollout/vllm_omni_async_server.py` — guard `rollout_attn_backend` access (diffusion-only field) so AR rollouts don't crash.
- `examples/gspo_trainer/config/qwen3_omni_thinker_gspo.yaml`, `tests/special_e2e/run_gspo_qwen3_omni_thinker_lora_smoke.sh` — explicit `target_modules`; `use_orig_params` + `wrap_policy.min_num_params` for LoRA weight sync.
- `tests/gpu_smoke/run_gpu_smoke_tests.sh` — re-add the recipe as Test 8 (2 GPUs).
- `examples/gspo_trainer/README.md` — document 4.x / 5.x support, `verl @ main`, `TransferQueue 0.1.8`.

### Checklist Before Submitting

- [x] Read the Contribute Guide.
- [x] Apply pre-commit checks: `pre-commit run` passes on changed files.
- [x] Add / Update the documentation (README).
- [x] Add end-to-end test to the CI workflow (gpu_smoke Test 8).

### CODE REVIEW

**Comment by qinganrice** [comment]
PTAL @SamitHuang @zhtmike @wtomin @knlnguyen1802 Thanks!

File: examples/gspo_trainer/README.md [hunk 1]
```diff
@@ -16,11 +16,12 @@ following stack (rollout↔actor pearson ≈ 0.993):
 | --- | --- |
 | vLLM | `0.22.0` |
 | vLLM-Omni | `0.22.0` (the `v0.22.0` release tag) |
-| transformers | `4.57.6` (see warning below) |
+| transformers | `4.57.6` (4.x) or `5.x` — see note below |
```
Comment (SamitHuang) [inline examples/gspo_trainer/README.md]: Drop 4.x compatibility for cleanness?

File: examples/gspo_trainer/README.md [hunk 2]
```diff
@@ -16,11 +16,12 @@ following stack (rollout↔actor pearson ≈ 0.993):
...
 | torch | `2.11.0+cu130` |
 | flash-attn | `2.8.3` |
 | accelerate | `1.12.0` |
-| verl | commit: `8a69493027` (w/ FSDP LoRA fixes) |
+| verl | `main` (w/ FSDP LoRA fixes) |
```
Comment (SamitHuang) [inline examples/gspo_trainer/README.md]: Specify the detailed commit id

File: examples/gspo_trainer/README.md [hunk 3]
```diff
@@ -34,17 +35,20 @@ pip install "vllm-omni @ git+https://github.com/vllm-project/vllm-omni.git@v0.22
 pip install pydub
 
 # verl (must include the FSDP layered-summon fix)
-pip install "verl @ git+https://github.com/verl-project/verl.git@8a69493027"
+pip install "verl @ git+https://github.com/verl-project/verl.git@main"
```
Comment (SamitHuang) [inline examples/gspo_trainer/README.md]: ditto

File: examples/gspo_trainer/README.md [hunk 4]
```diff
@@ -34,17 +35,20 @@ pip install "vllm-omni @ git+https://github.com/vllm-project/vllm-omni.git@v0.22
...
+
+# verl main's V1 trainer requires TransferQueue (not always pulled transitively)
+pip install TransferQueue==0.1.8
```
Comment (SamitHuang) [inline examples/gspo_trainer/README.md:42]: is TQ used in training?

**Review by SamitHuang — COMMENTED** [review]
Thanks for the speedy work. Pls attach the training results and convergene curve.

File: verl_omni/models/transformers/qwen3_omni_thinker.py [hunk 5]
```diff
@@ -167,10 +139,136 @@ def _patched_hf_processor(name_or_path, **kwargs):
...
+            if model is not None and type(model).__name__ == "Qwen3OmniMoeForConditionalGeneration":
+                return  # experts already unfused; tf5 name remapping must not run
+            return _orig_convert(peft_config, model=model, conversions=conversions)
+
+        _twc.get_model_conversion_mapping = _patched_get_mapping
+        _twc.convert_peft_config_for_transformers = _patched_convert
+    except (ImportError, AttributeError):
+        pass
+
+    def _patched_get_peft_model(model, peft_config, **kwargs):
```
Comment (zhtmike) [inline verl_omni/models/transformers/qwen3_omni_thinker.py:253]: Do we really need this?

Comment (qinganrice) [reply verl_omni/models/transformers/qwen3_omni_thinker.py:253]: Yes. On tf5, PEFT 0.19.1's `convert_peft_config_for_transformers` remaps `gate_proj/up_proj` → the fused `gate_up_proj `name. After we've unfused the experts into per-expert `nn.Linear,` that remapping rewrites our target_modules away from `gate_proj/up_proj/down_proj`, so LoRA never attaches to the experts (48 expert adapters → 0). I have added a clearer inline comment there.

File: examples/gspo_trainer/README.md (hunk 4 above)
Comment (qinganrice) [reply examples/gspo_trainer/README.md:42]: Yes, it is needed by verl side. I added some comments in README about this.

File: examples/gspo_trainer/README.md (hunk 3 above)
Comment (qinganrice) [reply examples/gspo_trainer/README.md]: Fix it

File: examples/gspo_trainer/README.md (hunk 2 above)
Comment (qinganrice) [reply examples/gspo_trainer/README.md]: add it

File: examples/gspo_trainer/README.md (hunk 1 above)
Comment (qinganrice) [reply examples/gspo_trainer/README.md]: Dropped

**Comment by qinganrice** [comment]
> Thanks for the speedy work. Pls attach the training results and convergene curve.

Already updated reward curve and performance stats in README and also attached it in PR description.

**Review by zhtmike — APPROVED** [review]

File: tests/special_e2e/run_gspo_qwen3_omni_thinker_lora_smoke.sh
Comment (wtomin) [inline tests/special_e2e/run_gspo_qwen3_omni_thinker_lora_smoke.sh:67]: This GPU smoke test runs in a CI env, and the env configuration file is `.github/workflows/gpu_smoke.yml`. This yaml file does not include the specific packages needed by qwen3-omni, like transformers `5.x` and `TransferQueue`. You should install them seperately in `tests/special_e2e/run_gspo_qwen3_omni_thinker_lora_smoke.sh`.

Comment (qinganrice) [reply tests/special_e2e/run_gspo_qwen3_omni_thinker_lora_smoke.sh:67]: Added.

**Review by wtomin — APPROVED** [review]
